### ISSUE TITLE
Document and test Managed Lane Workflow

### ISSUE THREAD

**Issue by e-lo** [wsp-sag/Lasso#60]
Four workflows to test and document:

- Coding a new one inline.
- Updating one inline.
- Updating one in parallel.
- Coding a new one parallel.

**Comment by jhelsel11** [comment]
@e-lo,

Can we have a phone call tomorrow to discuss? I am free anytime but 11:30 tomorrow morning. (It might be best to set up a quick gotomeeting so that I can show my questions on the screen.)

My most pressing question is that I know that Network Wrangler has an example managed lane card, but I don't know what the cube log file should look like to create that yml. The example yml has attributes like ML_Lanes and ML_Price which are not in the base network attributes (and so weren't brought into the cube network I'm trying to code.) Is the user expected to know these field names and add them during a log file? If I add those fields in a log file, then it will print a record for every link, even the ones that will be blank and that seems unnecessarily repetitive. (Also, the cube log file will output all the changes as ALL_CAPS, will that be a problem for these fields?)

I'm also not sure how the parallel coding is supposed to work. My understanding is that Cube will only permit one link to connect each a-b nodes and will throw an error if the links are not unique. Is that right? (And if Cube is willing to permit parallel links between a and b nodes, how do I pass that information on to the true shape display? Rachel has been very adamant that she wants to be able to edit in true shape, so I want to make sure I can explain the process to her.)

**Comment by jhelsel11** [comment]
(I'm also free this afternoon if you're free. You can reach me at 404.759.8844 or we can set up a gotomeeting.)

**Comment by jhelsel11** [comment]
Also, is there a way to get the yml to print with blank fields for the expected metadata? It seems to just start with a "changes" tag, but not the project, tags, dependencies fields.

**Comment by e-lo** [comment]
> the cube log file will output all the changes as ALL_CAPS, will that be a problem for these fields?

> If I add those fields in a log file, then it will print a record for every link

--> Lasso should take care of both of these things for you, but this is a good thing to confirm in your current work flow.

**Comment by jhelsel11** [comment]
Do I add a field called ml_lanes in Cube? If so, do I need to save the change for the entire network?

**Comment by e-lo** [comment]
> The example yml has attributes like ML_Lanes and ML_Price which are not in the base network attributes. Is the user expected to know these field names and add them during a log file?

--> Yes. Although thinking about a way to make this easier for MetCouncil would be to add them to the base network as null values.

cc: @i-am-sijia

**Comment by jhelsel11** [comment]
What do you think of this principle?
The base network should contain a field (even if blank) for every value the user might be expected to manipulate in a project card.

**Comment by e-lo** [comment]
> I'm also not sure how the parallel coding is supposed to work.

--> Are you concerned for the separated peak-direction facility (which has its own AB nodes)? or for the ones where you are coding the `ml_lanes`?  For the lanes on the same facility, Wrangler will take care of that and create new nodes, connected by "dummy links" that are a scalar value more than the existing ones so cube should be fine.

**Comment by jhelsel11** [comment]
The ones I'm coding with `ml_lanes`. I've got a project card made for the reversible lanes so I think that's ok. 
How does wrangler handle the geometry of the new links?

**Comment by i-am-sijia** [comment]
@e-lo - I saw ad-hoc variables can be allowed for link addition, but don't see them available for link changes.   Either we add the ad-hoc condition in link change, or we add all ad-hoc's to the base network as nulls.

**Comment by jhelsel11** [comment]
As the end user, I think it would be better to have the ad-hoc's in the base network (if I'm understanding this correctly), because then the user will not need (or be tempted) to save project cards with changes to the entire network after adding a new field.)

**Comment by e-lo** [comment]
> How does wrangler handle the geometry of the new links?

Since this is for the lanes that are technically part of the same facility, it duplicates the geometry and then offsets it.

**Comment by jhelsel11** [comment]
Do you have an example of a .log file with the kind of change you have in mind?

**Comment by e-lo** [comment]
@i-am-sijia : Since we *know* that there will be managed lanes (rather than guessing), I think it makes sense to add these particular variables to the base network.

**Comment by e-lo** [comment]
> ...the user will not need (or be tempted) to save project cards with changes to the entire network after adding a new field

I don't think it is necessary to add *all potential* link fields. It is fine if a user wants to add a new field...this is an expected use case.

**Comment by jhelsel11** [comment]
Ok. I'll try creating a project card that adds the field until I get a new base network.

**Comment by i-am-sijia** [comment]
@e-lo Ok I will add them to the base network.  Just ML_lanes and ML_price?  Wiki needs to be updated.

Also, I saw you added ad-hoc to link addition, but not link change. If we want to allow users to add a new field in the future to an existing link, I think we need to allow ad-hoc in link change as well.    am I right?

**Comment by jhelsel11** [comment]
@e-lo 
Can you confirm that these are the changes I would make?
For a link with a single ML during the AM peak that is otherwise open to the public (the parallel lane, not a reversible), I would increase the MD, PM, and NT lanes by 1 and then say ML Lanes = 1? Or do I need to create the lane project card and then go back and re-edit?

**Comment by e-lo** [comment]
The project card would be something like this:
```
property: lanes_am
change: -1
property: ml_lanes
set: 1
```
So editing the network you would reduce the lanes_am by 1 and set the ml_lanes = 1

**Comment by e-lo** [comment]
> Just ML_lanes and ML_price? and by time periods and category or not?

Should be everything in the [output variables list in Parameters](https://github.com/wsp-sag/Lasso/blob/master/lasso/parameters.py)

**Comment by jhelsel11** [comment]
If ML_lanes = 1, then will that give a managed lane for all time periods (and the am is alone in losing a general purpose lane) or will lasso interpret that to mean that the am gets a managed lane and not the other time periods?

**Comment by i-am-sijia** [comment]
I'm leaning toward thinking it should be ML_lanes_AM, ML_lanes_MD, ML_lanes_PM, ML_lanes_NT, and ML_price_sov_AM, ML_price_hov2_AM, etc.

**Comment by jhelsel11** [comment]
That's fine with me. Also, we have several price values in the network, e.g. price_sov_am and price_hov_pm. Is there any reason we wouldn't just use those for the prices? (Don't forget about the column name character limits.)

**Comment by jhelsel11** [comment]
@e-lo, how do prices work? Are they per mile? Given the fee schedule for mndot (https://www.dot.state.mn.us/mnpass/mnpassexpresslanes.html), what should I list as the default price for all of the managed lanes I'm building?

**Comment by e-lo** [comment]
Sorry, just want to write this down.  Right now the "standard network" definition doesn't define what prices represent. I would think that they would be the price for traversing a specific link for a specific user class (i.e. hov3) at a specific time of day (i.e. AM Peak).  I

I think the real question is what does the MetCouncil model expect the units to be in?  If it i s expecting anything other than the link cost. (i.e. per mile cost) then we should do a conversion in Lasso

**Comment by jhelsel11** [comment]
MetCouncil expects the number to be 0. They have a "toll choice" step in assignment where the prices are updated between model iterations.
